models_test.go

```diff
@@ -155,3 +155,24 @@ func (bc *BadComment) JSONAPILinks() *Links {
 		"self": []string{"invalid", "should error"},
 	}
 }
+
+type Company struct {
+	ID        string    `jsonapi:"primary,companies"`
+	Name      string    `jsonapi:"attr,name"`
+	Boss      Employee  `jsonapi:"attr,boss"`
+	Teams     []Team    `jsonapi:"attr,teams"`
+	FoundedAt time.Time `jsonapi:"attr,founded-at,iso8601"`
+}
+
+type Team struct {
+	Name    string     `json:"name"`
```

Comment:
I've read the [comment ](https://github.com/google/jsonapi/pull/99#issuecomment-320225690)@dev-mush left.

I think all these tags (even for a nested resource) should all be `jsonapi` rather than `json`. Why? What if you want to support serializations for both JSON and JSONAPI; you'd want separate tags for each. I'd recommend just using `jsonapi:"attr,names"` even for these nested struct's fields.

Reply:
@aren55555 according to the [spec](http://jsonapi.org/format/#document-resource-object-attributes) attributes need to be valid json but can't be a json:api object since they are not allowed to have links or relationships. 

Maybe I get this totaly wrong but this would allow the following:

```
{
	"data": {
		"id": "1324",
		"type": "article",
		"attributes": {
			"title": "The best article",
			"author": {
				"id": "18",
				"type": "author",
				"attributes": {
					"name": "Slemgrim"
				}
			}
		}
	}
}
```
where you also could do this:
```

{
	"data": {
		"id": "1324",
		"type": "article",
		"attributes": {
			"title": "The best article",
		}
	},
	"relationships": {
	    "author": {
	    	"data": {"id": "18", "type": "author"}
	    }
	},
	"included": [{
		"id": "18",
		"type": "author",
		"attributes": {
			"name": "Slemgrim"
		}
	}]
}
```

Reply:
Could the unmarshal/marshal not just return an error if anything other than an attr is serialized? Or it could ignore any tag that was not an `attr`?

Reply:
@Slemgrim bump on question from @aren55555 .  If we can resolve this, we'll merge this immediately.

Reply:
@aren55555 @shwoodard I tried to look into this today. The changes required to allow "jsonapi" marshalling for nested attributes, seem non-trivial with the current architecture. At least non-trivial to me.

I'm happy to spend my time on this if someone could give me an example where this would be needed. 

The jsonapi spec does not allow the fields "id, type, relationship, links" as attribute names. In other words: it doesn't allow to have a resource inside a resource. Allowing to tag the attributes of a non jsonapi resource with jsonapi tags would be missleading. It's like saying: "hey, you can name this json value jsonapi even though it will always be a normal json"

If you insist on this without further explanation then maybe one of the other contributers of this PR are willing to implement this. (@msabramo)

Reply:
Could the solution be to allow validate that id, type, and links are missing from nested structs?  Maybe we should allow relationships and process the data to be in the "included" part of the response?

Reply:
I agree with @aren55555 that we should use jsonapi for all properties serialized or deserialized by this lib.